File: spring-dao/src/main/java/org/r3al/springdao/DaoQueryBeanDefinition.java
package org.r3al.springdao;

import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.beans.factory.support.AbstractBeanDefinition;
import org.springframework.beans.factory.support.BeanDefinitionBuilder;

public class DaoQueryBeanDefinition {

    private DaoQueryBeanDefinition() {}

    public static AbstractBeanDefinition of(Class<? extends DaoQuery> classe, Object source) {
        BeanDefinitionBuilder builder = BeanDefinitionBuilder.rootBeanDefinition(classe.getName());
        builder.getRawBeanDefinition().setSource(source);
        builder.setLazyInit(false);
        builder.setScope(BeanDefinition.SCOPE_PROTOTYPE);
        AbstractBeanDefinition beanDefinition = builder.getBeanDefinition();
        beanDefinition.setInstanceSupplier(() -> source);
        beanDefinition.setAttribute("factoryBeanObjectType", classe.getName());
        return beanDefinition;
    }

}

File: spring-dao/src/main/kotlin/org/r3al/springdao/DaoQueryBeanDefinition.kt
package org.r3al.springdao

import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.beans.factory.support.AbstractBeanDefinition
import org.springframework.beans.factory.support.BeanDefinitionBuilder
import java.util.function.Supplier

internal object DaoQueryBeanDefinition {
    fun of(classe: Class<out DaoQuery>, source: Any?): AbstractBeanDefinition {
        val builder = BeanDefinitionBuilder.rootBeanDefinition(classe.name)
        builder.rawBeanDefinition.source = source
        builder.setLazyInit(false)
        builder.setScope(BeanDefinition.SCOPE_SINGLETON)
        val beanDefinition = builder.beanDefinition
        beanDefinition.instanceSupplier = Supplier { source }
        beanDefinition.setAttribute("factoryBeanObjectType", classe.name)
        return beanDefinition
    }
}